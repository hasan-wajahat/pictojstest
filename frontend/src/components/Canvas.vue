<template>
  <div class="canvas col-sm-8 col-md-8 col-lg-8">
    <div class="block canvas" @dragover="onDragOver" @drop="onDrop">
      <template v-for="(asset, index) in assets">
        <div
          draggable="true"
          @dragstart="onDragStart(asset, $event)"
          @drop="onItemDrop(asset, $event)"
          @dragover="onDragOver"
          v-if="asset.type === 'image'"
          :key="asset.id"
          class="image"
          :style="getStyles(index)"
        >
          <img :src="asset.url">
          <button @click="$emit('delete-asset', index)" class="delete-button btn-danger">
            <i class="fa fa-close"></i>
          </button>
        </div>
      </template>
    </div>
  </div>
</template>

<script>
export default {
  name: "Canvas",
  props: {
    assets: Array
  },
  data: function() {
    return {
      movedObjectSize: {
        x: 0,
        y: 0
      }
    };
  },
  methods: {
    onDragStart: function(asset, event) {
      event.dataTransfer.setData("text", asset.id);
      this.movedObjectSize = {
        x: event.target.clientWidth,
        y: event.target.clientHeight
      };
    },
    onDragOver: function(event) {
      event.preventDefault();
    },
    onDrop: function(event) {
      event.preventDefault();
      const id = Number(event.dataTransfer.getData("text"));
      let offsetValues = {
        left: event.offsetX - this.movedObjectSize.x / 2,
        top: event.offsetY - this.movedObjectSize.y / 2
      };
      this.$emit("asset-moved", { id, offsetValues });
    },
    onItemDrop: function(asset, event) {
      event.preventDefault();
      event.stopPropagation();
      const id = Number(event.dataTransfer.getData("text"));
      let offsetValues = {
        left: asset.left + event.offsetX - this.movedObjectSize.x / 2,
        top: asset.top + event.offsetY - this.movedObjectSize.y / 2
      };
      this.$emit("asset-moved", { id, offsetValues });
    },
    getStyles: function(index) {
      return {
        left: `${this.assets[index].left}px`,
        top: `${this.assets[index].top}px`
      };
    }
  }
};
</script>

<style lang="scss" scoped>
.canvas {
  position: relative;
  overflow: hidden;
}
.image {
  cursor: pointer;
  position: absolute;
  img {
    max-width: 90%;
  }
  &:hover {
    background: darken(white, 5%);
    .delete-button {
      display: block;
    }
  }
}
.delete-button {
  padding: 2px 4px;
  position: absolute;
  top: -5px;
  right: -5px;
  font-size: 12px;
  border-radius: 50%;
  cursor: pointer;
  display: none;
}
</style>