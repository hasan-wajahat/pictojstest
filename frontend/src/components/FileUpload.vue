<template>
  <div class="form">
    <h3>Form</h3>
    <input
      type="file"
      id="file-input"
      class="form-control"
      placeholder="Upload Your Images"
      accept="image/*"
      @change="filesChange($event);"
    >
    <button @click="onClick" class="btn btn-default">upload</button>
  </div>
</template>

<script>
import { uploadFile } from "../helpers/Api";
export default {
  name: "FileUpload",
  data: function() {
    return {
      file: null,
    };
  },
  methods: {
    filesChange: function(event) {
      this.file = event.target.files[0];
    },
    onClick: async function() {
      try {
        const response = await uploadFile('/uploads', this.file);
        if (response) {
          this.$emit('file-upload', response.file);
          document.getElementById("file-input").value = "";

        }
      } catch (error) {
        // eslint-disable-next-line
        console.log(Error);
      }
    }
  }
};
</script>