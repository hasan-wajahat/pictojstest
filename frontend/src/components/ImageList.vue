<template>
  <div class="image">
    <h4>Images</h4>
    <ul class="list-unstyled">
      <li
        :key="index"
        v-for="(image, index) in images"
        class="image-asset"
        @click="$emit('asset-click', {url: image, type: 'image'})"
      >
        <img :src="image" class="img-rounded non-draggable">
      </li>
    </ul>
  </div>
</template>


<script>
export default {
  name: "ImageList",
  props: {
    images: Array
  }
};
</script>

<style lang="scss" scoped>
.image-asset {
  cursor: pointer;
  transition: transform 0.2s;
  &:hover {
    transform: scale(1.1);
  }
}
.non-draggable {
  user-drag: none;
  user-select: none;
  -moz-user-select: none;
  -webkit-user-drag: none;
  -webkit-user-select: none;
  -ms-user-select: none;
}
</style>