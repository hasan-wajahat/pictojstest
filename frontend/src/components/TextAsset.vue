<template>
  <div class="text">
    <h4>Text</h4>
    <button @click="$emit('add-text', {type: 'text', text: ''})" class="btn btn-default">Add Text</button>
  </div>
</template>

<script>
export default {
  name: "TextAsset"
};
</script>