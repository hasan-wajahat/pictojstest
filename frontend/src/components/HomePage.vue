<template>
  <div class="main-container">
    <div class="sidepane col-sm-2 col-md-2 col-lg-2">
      <FileUpload v-on:file-upload="onFileUpload"/>
      <hr>
      <div class="assets">
        <h3>Assets</h3>
        <TextAsset v-on:add-text="onAddAsset" />
        <ImageList :images="images" v-on:asset-click="onAddAsset"/>
      </div>
    </div>
    <Canvas :assets="assets" v-on:asset-moved="onAssetMoved" v-on:delete-asset="onDeleteAsset"/>
  </div>
</template>

<script>
import { getData } from "../helpers/Api";
import ImageList from "./ImageList";
import FileUpload from "./FileUpload";
import Canvas from "./Canvas";
import TextAsset from "./TextAsset";

export default {
  name: "HomePage",
  data: function() {
    return {
      images: [],
      assets: []
    };
  },
  components: {
    ImageList,
    FileUpload,
    Canvas,
    TextAsset,
  },
  mounted: async function() {
    try {
      const response = await getData("/images");
      this.images = response;
    } catch (error) {
      // eslint-disable-next-line
      console.log(error);
    }
  },
  methods: {
    onFileUpload: function(url) {
      this.images.push(url);
    },
    onAddAsset: function(asset) {
      this.assets.push({
        ...asset,
        id: Date.now(),
        left: 0,
        top: 0
      });
    },
    onAssetMoved: function({ id, offsetValues }) {
      const index = this.assets.findIndex(asset => asset.id === id);
      const currentAsset = this.assets[index];
      const newAsset = {
        ...currentAsset,
        ...offsetValues
      };
      this.assets.splice(index, 1, newAsset);
    },
    onDeleteAsset: function(index) {
      this.assets.splice(index, 1);
    }
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style lang="scss" scoped>
.main-container {
  height: 100%;
}
</style>
