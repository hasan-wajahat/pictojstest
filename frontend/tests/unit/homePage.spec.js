import { shallowMount } from '@vue/test-utils'
import HomePage from '@/components/HomePage.vue'
import ImageList from "@/components/ImageList";
import FileUpload from "@/components/FileUpload";
import Canvas from "@/components/Canvas";
import TextAsset from "@/components/TextAsset";

// hack method to mock fetch
// need to fix mock
global.fetch = () => Promise.resolve({
    json: () => {}
})

describe('HomePage.vue', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(HomePage);
  })

  it('renders container', () => {
    expect(wrapper.find('div')).toBeTruthy();
  })

  it('renders all components', () => {
    expect(wrapper.find(ImageList)).toBeTruthy();
    expect(wrapper.find(FileUpload)).toBeTruthy();
    expect(wrapper.find(Canvas)).toBeTruthy();
    expect(wrapper.find(TextAsset)).toBeTruthy();
  })

  it('file upload adds to images', () => {
    const testImage = 'test.png';
    wrapper.find(FileUpload).vm.$emit('file-upload', testImage);
    expect(wrapper.vm.images).toEqual([testImage]);
  })

  it('can add text assets', () => {
    const testAsset = {text: '', id: 1};
    wrapper.find(TextAsset).vm.$emit('add-text', testAsset);
    expect(wrapper.vm.assets).toMatchSnapshot();
  })

  it('can add image assets', () => {
    const testAsset = {url: 'test.png', id: 2};
    wrapper.find(ImageList).vm.$emit('asset-click', testAsset);
    expect(wrapper.vm.assets).toMatchSnapshot();
  })

  describe('asset added', () => {
    beforeEach(() => {
      const testAsset = {text: '', id: 1};
      wrapper.find(TextAsset).vm.$emit('add-text', testAsset);
    })

    it('updates values when asset is moved', () => {
      const update = {
        id: 1, 
        offsetValues: {left: 10, top: 10}
      }
      wrapper.find(Canvas).vm.$emit('asset-moved', update);
      expect(wrapper.vm.assets).toMatchSnapshot();
    })

    it('can delete asset', () => {
      wrapper.find(Canvas).vm.$emit('delete-asset', 0);
      expect(wrapper.vm.assets).toEqual([]);
    })
  })
})
