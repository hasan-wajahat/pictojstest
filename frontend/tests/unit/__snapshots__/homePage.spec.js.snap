// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage.vue asset added updates values when asset is moved 1`] = `
Array [
  Object {
    "id": 1,
    "left": 10,
    "text": "",
    "top": 10,
  },
]
`;

exports[`HomePage.vue can add image assets 1`] = `
Array [
  Object {
    "id": 2,
    "left": 0,
    "top": 0,
    "url": "test.png",
  },
]
`;

exports[`HomePage.vue can add text assets 1`] = `
Array [
  Object {
    "id": 1,
    "left": 0,
    "text": "",
    "top": 0,
  },
]
`;
